openstack_controller/datadog_checks/openstack_controller/api/api_sdk.py

```diff
@@ -0,0 +1,311 @@
+# (C) Datadog, Inc. 2023-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from os import environ
+
+from keystoneauth1 import session
+from keystoneauth1.identity import v3
+from openstack import connection
+from openstack.config import loader
+
+from datadog_checks.openstack_controller.api.api import Api
+from datadog_checks.openstack_controller.api.catalog import Catalog
+from datadog_checks.openstack_controller.components.component import Component
+
+
+class ApiSdk(Api):
+    def __init__(self, config, logger, http):
+        super(ApiSdk, self).__init__()
+        self.log = logger
+        self.config = config
+        self.http = http
+        self._add_microversion_headers()
+
+        # Set the environment variable to the path of the config file for openstacksdk to find it
+        environ["OS_CLIENT_CONFIG_FILE"] = self.config.openstack_config_file_path
+        self.cloud_config = loader.OpenStackConfig(config_files=[self.config.openstack_config_file_path]).get_one_cloud(
+            cloud=self.config.openstack_cloud_name
+        )
+        self._interface = (
+            self.config.endpoint_interface
+            if self.config.endpoint_interface
+            else self.cloud_config.get_auth_args().get('interface', 'public')
+        )
+        self._region_id = (
+            self.config.endpoint_region_id
+            if self.config.endpoint_region_id
+            else self.cloud_config.get_auth_args().get('region_name')
+        )
+
+        self.connection = None
+        self._access = None
+        self._catalog = None
+
+    def _add_microversion_headers(self):
+        if self.config.nova_microversion:
+            self.log.debug("adding X-OpenStack-Nova-API-Version header to `%s`", self.config.nova_microversion)
+            self.http.options['headers']['X-OpenStack-Nova-API-Version'] = self.config.nova_microversion
+
+        if self.config.ironic_microversion:
+            self.log.debug("adding X-OpenStack-Ironic-API-Version header to `%s`", self.config.ironic_microversion)
+            self.http.options['headers']['X-OpenStack-Ironic-API-Version'] = self.config.ironic_microversion
+
+    def auth_url(self):
+        return self.cloud_config.get_auth_args().get('auth_url')
+
+    def has_admin_role(self):
+        self.log.debug("role_names: %s", self._access.role_names)
+        return 'admin' in self._access.role_names
+
+    def component_in_catalog(self, component_types):
+        return self._catalog.has_component(component_types)
+
+    def authorize_user(self):
+        v3_auth = v3.Password(
+            auth_url=self.cloud_config.get_auth_args().get('auth_url'),
+            username=self.cloud_config.get_auth_args().get('username'),
```

Comment:
What do you mean with constants? They come from something variable as it is the content of a user's file

Reply:
Hm I was thinking something like setting a constant `SDK_AUTH_CONFIG='auth_url'` since we use the key in a few places but I don't feel strongly either way